core/staking_verifier_test.go

```diff
@@ -0,0 +1,149 @@
+		SecurityContact: "LicenseToKill",
+		Details:         "blah blah blah",
+	}
+	rate, _ := numeric.NewDecFromStr("0.1")
+	maxRate, _ := numeric.NewDecFromStr("0.5")
+	maxChangeRate, _ := numeric.NewDecFromStr("0.05")
+	commission := staking.CommissionRates{
+		Rate:          rate,
+		MaxRate:       maxRate,
+		MaxChangeRate: maxChangeRate,
+	}
+	minSelfDel := big.NewInt(1e18)
+	maxTotalDel := big.NewInt(9e18)
+	pubKey, pubSig := generateBlsKeySigPair()
+	slotPubKeys := []shard.BlsPublicKey{pubKey}
+	slotKeySigs := []shard.BLSSignature{pubSig}
+	amount := big.NewInt(5e18)
+	v := staking.CreateValidator{
+		ValidatorAddress:   validatorAddress,
+		Description:        desc,
+		CommissionRates:    commission,
+		MinSelfDelegation:  minSelfDel,
+		MaxTotalDelegation: maxTotalDel,
+		SlotPubKeys:        slotPubKeys,
+		SlotKeySigs:        slotKeySigs,
+		Amount:             amount,
+	}
+	return &v
+}
+
+// Test CV1: create validator
+func Test_CV1(t *testing.T) {
+	statedb, _ := state.New(common.Hash{}, state.NewDatabase(ethdb.NewMemDatabase()))
+	msg := createValidator()
+	statedb.AddBalance(msg.ValidatorAddress, big.NewInt(5e18))
+	if _, err := VerifyAndCreateValidatorFromMsg(statedb, big.NewInt(0), big.NewInt(0), msg); err != nil {
+		t.Error("expected", nil, "got", err)
+	}
+}
+
+// Test CV3: validator already exists
+func Test_CV3(t *testing.T) {
+	statedb, _ := state.New(common.Hash{}, state.NewDatabase(ethdb.NewMemDatabase()))
+	msg := createValidator()
+	statedb.AddBalance(msg.ValidatorAddress, big.NewInt(5e18))
+	if _, err := VerifyAndCreateValidatorFromMsg(statedb, big.NewInt(0), big.NewInt(0), msg); err != nil {
+		t.Error("expected", nil, "got", err)
+	}
+	statedb.SetValidatorFlag(msg.ValidatorAddress)
+
+	if _, err := VerifyAndCreateValidatorFromMsg(statedb, big.NewInt(0), big.NewInt(0), msg); !strings.Contains(err.Error(), errValidatorExist.Error()) {
+		t.Error("expected", errValidatorExist, "got", err)
+	}
+}
+
+// Test CV5: identity longer than 140 characters
+func Test_CV5(t *testing.T) {
+	statedb, _ := state.New(common.Hash{}, state.NewDatabase(ethdb.NewMemDatabase()))
+	msg := createValidator()
+	// identity length: 200 characters
+	msg.Identity = "adsfwryuiwfhwifbwfbcerghveugbviuscbhwiefbcusidbcifwefhgciwefherhbfiwuehfciwiuebfcuyiewfhwieufwiweifhcwefhwefhwiewwerfhuwefiuewfhuewhfiuewhfefhshfrhfhifhwbfvberhbvihfwuoefhusioehfeuwiafhaiobcfwfhceirui"
+	statedb.AddBalance(msg.ValidatorAddress, big.NewInt(5e18))
+	identitylengthError := ctxerror.New("[EnsureLength] Exceed Maximum Length", "have", len(msg.Identity), "maxIdentityLen", staking.MaxIdentityLength)
+	if _, err := VerifyAndCreateValidatorFromMsg(statedb, big.NewInt(0), big.NewInt(0), msg); err.Error() != identitylengthError.Error() {
+		t.Error("expected", identitylengthError, "got", err)
+	}
+}
+
+// Test CV6: website longer than 140 characters
+func Test_CV6(t *testing.T) {
+	statedb, _ := state.New(common.Hash{}, state.NewDatabase(ethdb.NewMemDatabase()))
```

Comment:
This is repeated over and over , write a function to create state and do the initial validator

Reply:
I have told with ganesha about this issue. he thought these tasks could run parallel, so I need to create statedb and validator for every test to prevent concurrency things, even it looks like redundant.

Reply:
if we want to run the tests in parallel and make the tests self contained, it is better to create the statedb freshly. @Renewwen you could create state once and do Reset whenever you need fresh state.